### PR TITLE
Refactor all the things

### PR DESCRIPTION
**Pull request by SouthEndMusic**
I plowed trough some refactoring ideas:
- Add the node type specific index to `NodeID`, which is used to avoid many lookups
- Precalculate: `FlowBoundary` outflow edges, `DiscreteControl` controlled nodes
- Split the `logic_mapping` per `DiscreteControl` node
- Let `DiscreteControl` read and write variables via refs. This neatly cleans up what was first partly node type specific and type unstable code. The source of truth for which parameters are controllable is now in the function `controllablefields` in `validation.jl`, which can also be used to automatically generate a table of controllable parameters in the docs

### CODE REVIEW

**Comment by evetion** [comment]
Was your secondary goal to touch all Julia core files? 😅

File: core/src/validation.jl [hunk 1]
```diff
@@ -86,6 +86,17 @@ n_neighbor_bounds_control(::Val{:FlowDemand}) = n_neighbor_bounds(0, 0, 1, 1)
 n_neighbor_bounds_control(nodetype) =
     error("'n_neighbor_bounds_control' not defined for $nodetype.")
 
+controllablefields(nodetype::Symbol) = controllablefields(Val(nodetype))
```
Comment (evetion) [inline core/src/validation.jl:89]: Could also be placed directly underneath the definition of each struct?

**Review by evetion — APPROVED** [review]
Overall LGTM. I didn't check the tests, and it might be good if @visr could check the code in `util.jl`, as most of the domain changes are in there. The other changes are mostly related to using/replacing the NodeId.idx.

**Comment by evetion** [comment]
AGV model went from:

`2.030423 seconds (3.86 M allocations: 1006.978 MiB, 1.50% gc time)`
to
`1.934420 seconds (3.02 M allocations: 966.206 MiB, 1.83% gc time)`

Given the reduction in allocations, this might be even faster on non-control heavy models.

File: core/src/parameter.jl [hunk 2]
```diff
@@ -16,14 +16,36 @@ NodeType.T(str::AbstractString) = NodeType.T(Symbol(str))
...
+end
+
+function NodeID(type::NodeType.T, value::Integer, db::DB)::NodeID
+    node_type_string = string(type)
+    idx = only(
+        only(
+            execute(
+                columntable,
+                db,
+                "SELECT COUNT(*) FROM Node WHERE node_type == $(esc_id(node_type_string)) AND node_id <= $value",
```
[the remark this answers is not in the document]
Comment (visr) [reply core/src/parameter.jl:37]: Added an inline comment. Why do you think this assumes order? It only counts, regardless of order.

File: core/src/validation.jl (hunk 1 above)
Comment (visr) [reply core/src/validation.jl:89]: Agree this isn't so much validation, but I feel like it fits better here than scattered across parameter.jl.

File: core/src/parameter.jl [hunk 3]
```diff
@@ -16,14 +16,36 @@ NodeType.T(str::AbstractString) = NodeType.T(Symbol(str))
 struct NodeID
     type::NodeType.T
     value::Int32
+    idx::Int32
```
[the remark this answers is not in the document]
Comment (visr) [reply core/src/parameter.jl:33]: Done in 7769ca4136849f65493d1c048d299f8f8638ef1d

**Pull request merged**